File: 005.后端工程师 (java工程师)/expense-tracker-server/src/main/java/com/xingzhewk/service/store/SmsCodeStore.java
package com.xingzhewk.service.store;

/**
 * 短信验证码存储。
 *
 * 抽象同 {@link LoginAttemptStore}：内存实现用于 dev/单机，
 * Redis 实现用于集群部署。所有实现需保证：
 *   - 写入时按 TTL 自动过期；
 *   - 支持「发送间隔」语义（距上次写入小于 minIntervalMs 时拒绝写入）。
 */
public interface SmsCodeStore {

    /**
     * 写入一条新验证码。若距上次写入小于 minIntervalMs，返回 false 表示频控拒绝。
     *
     * @param key           手机号
     * @param code          验证码
     * @param ttlSeconds    存活时长
     * @param minIntervalMs 最小发送间隔
     * @return true 写入成功；false 被频控拦截
     */
    boolean putIfNotThrottled(String key, String code, long ttlSeconds, long minIntervalMs);

    /**
     * 取验证码。已过期或不存在返回 null。
     */
    String get(String key);

    /**
     * 校验后删除（一次性）。
     */
    void remove(String key);
}

File: 003.前端代码/backend/store/smsCodeStore.js
/**
 * 短信验证码存储。
 *
 * 与 Java 后端 {@code com.xingzhewk.service.store.SmsCodeStore} 接口对齐。
 * 详见 contracts/DIFFS.md #2 #3。
 *
 * 仅内存实现（V1.1 不上 Redis，与 Java dev 默认一致）。
 */
class InMemorySmsCodeStore {
    constructor({ cleanupIntervalMs = 60 * 1000 } = {}) {
        this.store = new Map();
        this._cleanupTimer = setInterval(() => this.cleanupExpired(), cleanupIntervalMs);
        if (this._cleanupTimer.unref) this._cleanupTimer.unref();
    }

    /**
     * 写入一条新验证码。若距上次写入小于 minIntervalMs 返回 false（频控）。
     */
    putIfNotThrottled(key, code, ttlSeconds, minIntervalMs) {
        const now = Date.now();
        const prev = this.store.get(key);
        if (prev && now - prev.writeTimeMs < minIntervalMs) {
            return false;
        }
        this.store.set(key, {
            code,
            writeTimeMs: now,
            expireAtMs: now + ttlSeconds * 1000
        });
        return true;
    }

    /** 已过期或不存在返回 null。 */
    get(key) {
        const r = this.store.get(key);
        if (!r) return null;
        if (Date.now() > r.expireAtMs) {
            this.store.delete(key);
            return null;
        }
        return r.code;
    }

    /** 校验后删除（一次性消费）。 */
    remove(key) {
        this.store.delete(key);
    }

    cleanupExpired() {
        const now = Date.now();
        for (const [k, v] of this.store.entries()) {
            if (now > v.expireAtMs) this.store.delete(k);
        }
    }

    size() { return this.store.size; }

    destroy() {
        if (this._cleanupTimer) clearInterval(this._cleanupTimer);
        this._cleanupTimer = null;
    }
}

module.exports = { InMemorySmsCodeStore };
